src/workerd/jsg/modules-new-test.c++

```diff
@@ -0,0 +1,1889 @@
+    KJ_ASSERT(module.isMain());
+  }
+
+  {
+    // A built-in module is resolved when using a builtin contxt
+    auto& module = KJ_ASSERT_NONNULL(resolve(*registry, ResolveContext::Type::BUILTIN, bar));
+    KJ_ASSERT(module.specifier() == bar);
+    KJ_ASSERT(module.type() == Module::Type::BUILTIN);
+    KJ_ASSERT(module.isEsm());
+    KJ_ASSERT(!module.isMain());
+  }
+
+  {
+    // A built-in only module is resolved when using a built-in context
+    auto& module = KJ_ASSERT_NONNULL(resolve(*registry, ResolveContext::Type::BUILTIN, baz));
+    KJ_ASSERT(module.specifier() == baz);
+    KJ_ASSERT(module.type() == Module::Type::BUILTIN_ONLY);
+    KJ_ASSERT(!module.isEsm());
+    KJ_ASSERT(!module.isMain());
+  }
+
+  {
+    // A built-in only module is resolved when using a built-in only context
+    auto& module = KJ_ASSERT_NONNULL(resolve(*registry, ResolveContext::Type::BUILTIN_ONLY, baz));
+    KJ_ASSERT(module.specifier() == baz);
+    KJ_ASSERT(module.type() == Module::Type::BUILTIN_ONLY);
+    KJ_ASSERT(!module.isEsm());
+    KJ_ASSERT(!module.isMain());
+  }
+
+  // A built-in only module cannot be resolved from a bundle context
+  KJ_ASSERT(resolve(*registry, ResolveContext::Type::BUNDLE, baz) == kj::none);
+
+  // Fallback modules cannot be resolved from a built-in context
+  KJ_ASSERT(resolve(*registry, ResolveContext::Type::BUILTIN, foo) == kj::none);
+  KJ_ASSERT(resolve(*registry, ResolveContext::Type::BUILTIN_ONLY, foo) == kj::none);
+
+  // Bundle modules cannot be resolved from a built-in or built-in only context
+  KJ_ASSERT(resolve(*registry, ResolveContext::Type::BUILTIN, qux) == kj::none);
+  KJ_ASSERT(resolve(*registry, ResolveContext::Type::BUILTIN_ONLY, qux) == kj::none);
+
+  // We should have seen eleven distinct resolution events.
+  KJ_ASSERT(observer.modules.size() == 11);
+}
+
+// ======================================================================================
+
+KJ_TEST("Bundle shadows built-in") {
+  // A bundle module can shadow a built-in
+  ResolveObserverImpl observer;
+  ModuleRegistry::Builder registryBuilder(observer);
+
+  const auto foo = "foo:bar"_url;
+
+  ModuleBundle::BuiltinBuilder builtinBuilder;
+  auto source = "export const foo = 123;"_kjc;
+  builtinBuilder.addEsm(foo, source.asArray());
+  registryBuilder.add(builtinBuilder.finish());
+
+  ModuleBundle::BundleBuilder bundleBuilder;
+  kj::String bundleSource = kj::str("export const foo = 456;");
+  bundleBuilder.addEsmModule("foo:bar", bundleSource.releaseArray(), Module::Flags::MAIN);
+  registryBuilder.add(bundleBuilder.finish());
+
+  auto registry = registryBuilder.finish();
+
+  ResolveContext context {
+    .type = ResolveContext::Type::BUNDLE,
+    .source = ResolveContext::Source::OTHER,
+    .specifier = foo,
+    .referrer = ModuleBundle::BundleBuilder::BASE,
+  };
+
+  auto& module = KJ_ASSERT_NONNULL(registry->resolve(context));
+  KJ_ASSERT(module.specifier() == foo);
+  KJ_ASSERT(module.type() == Module::Type::BUNDLE);
```

Comment:
How do these asserts show that the built-in has been shadowed? Is it through the isMain() check?

Reply:
The `Module::Type::BUNDLE` check